tests/test_multi_connection.c

```diff
@@ -190,12 +192,245 @@ test_new(void **state)
+        /* Ignore these events as they are not on the handler thread */
+        return;
+    default:
+        break;
+    }
+
+    if (pvt_data->fail) {
+        /* was not expected to be called */
+        fail();
+    }
+}
+
+static int
+suspend_rpc_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *op_path, const sr_val_t *input, const size_t input_cnt,
+        sr_event_t event, uint32_t request_id, sr_val_t **output, size_t *output_cnt, void *private_data)
+{
+    struct cb_data *pvt_data = (struct cb_data *)private_data;
+
+    (void)session;
+    (void)sub_id;
+    (void)op_path;
+    (void)input;
+    (void)input_cnt;
+    (void)event;
+    (void)request_id;
+    (void)output;
+    (void)output_cnt;
+    if (pvt_data->fail) {
+        /* was not expected to be called */
+        fail();
+    }
+    return SR_ERR_OK;
+}
+
+static void
+test_sub_suspend(void **state)
+{
+    struct state *st = (struct state *)*state;
+    int i, j, ret;
+    sr_subscription_ctx_t *subscr[5] = {NULL};
+    struct cb_data pvt_data[2];
+    char xpath[128] = "";
+    const char *rpc_xpath[] = {
+        "/ops:rpc3",
+        "/ops:rpc2"
+    };
+
+    sr_val_t input, *output;
+    size_t output_count;
+
+    sr_data_t *data;
+
+    /* Create two subscriptions of each kind and suspend the second */
+    for (i = 1; i >= 0; i--) {
+        j = 0;
+        /* make the first one fail if invoked in a suspended state */
+        pvt_data[i].fail = i;
+
+        /* change_sub */
+        ret = sr_module_change_subscribe(st->sess1, "ietf-interfaces", NULL, module_change_cb, &pvt_data[i], 0, 0, &subscr[j]);
+        assert_int_equal(ret, SR_ERR_OK);
+        test_sub_suspend_helper(subscr[j], i);
+        j++;
+        /* oper get sub */
+        sprintf(xpath, "/ietf-interfaces:interfaces-state/interface[name='eth%d']/statistics", i);
+        ret = sr_oper_get_subscribe(st->sess1, "ietf-interfaces", xpath, suspend_oper_get_cb,
+                &pvt_data[i], 0, &subscr[j]);
+        assert_int_equal(ret, SR_ERR_OK);
+        test_sub_suspend_helper(subscr[j], i);
+        j++;
+
+        /* oper poll sub */
+        ret = sr_oper_poll_subscribe(st->sess1, "ietf-interfaces", xpath, 3000, 0, &subscr[j]);
+        assert_int_equal(ret, SR_ERR_OK);
+        test_sub_suspend_helper(subscr[j], i);
+        j++;
+
+        /* notification subscriptions */
+        ret = sr_notif_subscribe(st->sess1, "ops", NULL, 0, 0, suspend_notif_cb, &pvt_data[i], 0, &subscr[j]);
+        assert_int_equal(ret, SR_ERR_OK);
+        test_sub_suspend_helper(subscr[j], i);
+        j++;
+
+        /* RPC/action subscriptions */
+        ret = sr_rpc_subscribe(st->sess1, rpc_xpath[i], suspend_rpc_cb, &pvt_data[i], 0, 0, &subscr[j]);
+        assert_int_equal(ret, SR_ERR_OK);
+        test_sub_suspend_helper(subscr[j], i);
+    }
+
+    /* all subscriptions are set up, wait for threads to be spawned, and exercise them all */
+    usleep(100000);
+    sched_yield();
```

Comment:
Please do not use such constructions, they are never too reliable. Instead I usually create 2 threads and synchronize them using a barrier.

Reply:
Yes, you are right. I had put it in before when I had some threads, but now this is unnecessary and can be removed.
The subscriptions are already in ext SHM and publishers will have to wait for them to be invoked anyway.